### ISSUE TITLE
MiniEngine: High refresh rate = no particle rendering (ThreadGroupCountX = 0)

### ISSUE THREAD

**Issue by alessiot89** [microsoft/DirectX-Graphics-Samples#112]
File: ParticleEffect.cpp
Line: 146-147

```
    UINT NumSpawnThreads = (UINT)(m_EffectProperties.EmitRate * timeDelta);
    CompContext.Dispatch((NumSpawnThreads + 63) / 64, 1, 1);
```

When _timeDelta_ is to smaller then 1/64 (ie more then 64 frames per second) there is no dispatch since GroupCountX is set to 0.

If we change to

```
    UINT NumSpawnThreads = (UINT)(m_EffectProperties.EmitRate * timeDelta);
    CompContext.Dispatch((NumSpawnThreads + 64) / 64, 1, 1);
```

should work in every situation.

**Comment by stanard** [comment]
Hi,

Yes, this is a bug.  I deliberately left it unfixed for a while as I prepared a presentation about the tiled particle renderer.  I needed to have the same behavior so I could repeat performance measurements and have repeatable and consistent particle state.

Now that my presentation is finished, I will look into fixing this issue.  What needs to happen is

1) Compute fractional new particle count from EmitRate * timeDelta.  Save off fractional part so that it carries over to the next frame:  NewParticles = floor(EmitRate * timeDelta + Remainder).
2) Issue a UAV barrier between particle update and particle spawn so that new particles don't consume all available space forcing us to throw away existing particles.
3) Re-tune all of the effects so that their spawn rate and max particle count work together.

And my stretch goal is to replace the random number generator so that it happens on the GPU rather than on the CPU.  :-)

-James
